persistence/pg.go

```diff
@@ -378,7 +412,7 @@ func (p *PostgresPersister) finalizeInvoice(ctx context.Context,
 		return err
 	}
 	if result.RowsAffected() == 0 {
-		return types.ErrInvoiceNotFound
+		return ErrInvoiceAlreadyFinal
```

Comment:
Hmm, looking at this function in isolation, I am not sure about this change.

Maybe the update needs a where clause checking for `SettleRequested` too.

I suppose you can't distinguish between a non-existing invoice and an invoice that is already final.

Reply:
when calling `finalizeInvoice` we always do a SELECT FOR UPDATE, in the same transaction.

So, the only case where `result.RowsAffected() == 0` are:

- invoice already in invoiceFinalStatus state
- invoice was removed from the DB manually in the meantime

However, I think that it's important to add a Where clause `WHERE status = SETTLE_REQUESTED`. Because, IMO (even if it should never happen), `finalizeinvoice` should not changed the status of an invoice that has already reached it's final status.
(currently if invoice is marked as FAILED, and we call `finalizeinvoice(SETTLED)`, it will change the invoice final status)

I've added this where statement.
Did I miss the point? (if the goal is to really know if the invoice wasn't deleted in the meantime, we can re-do a SELECT statement. But we still have no guarantee, we only narrow the gap)

Reply:
Yes, I know that those outcomes are the only ones. But I mean that just looking at this function in isolation, the context with select-for-update isn't visible.

Maybe it is enough to just rename the error to something that is always true for this function. Which is inevitably something more generic as `ErrFinalizationFailed`. But then at least you can read this function without assumptions on the external environment.

Reply:
ahhh, didn;t understand that, will do!